xapian-letor/tests/api_letor.cc

```diff
@@ -0,0 +1,185 @@
+			     "massive beasts - I wouldn't want to meet a "
+			     "hungry one anywhere. Lions are scary even when "
+			     "lyin' down. Bears are scary even when bare. "
+			     "Together I suspect they'd be less scary, as the "
+			     "tigers, lions, and bears would all keep each "
+			     "other busy. On the other hand, bears don't live "
+			     "in the same continent as far as I know.");
+    db.add_document(doc);
+}
+
+DEFINE_TESTCASE(createfeaturevector, generated)
+{
+    Xapian::FeatureList fl;
+    Xapian::Database db = get_database("apitest_ranker1",
+				       db_index_two_documents);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("lions"));
+    Xapian::MSet mset;
+    vector<Xapian::FeatureVector> fv =
+	    fl.create_feature_vectors(mset, Xapian::Query("lions"), db);
+    TEST(fv.empty());
+    mset = enquire.get_mset(0, 10);
+    TEST(!mset.empty());
+    fv = fl.create_feature_vectors(mset, Xapian::Query("lions"), db);
+    TEST_EQUAL(fv.size(), 2);
+    TEST_EQUAL(fv[0].get_fcount(), 19);
+    TEST_EQUAL(fv[1].get_fcount(), 19);
+
+    return true;
+}
+
+DEFINE_TESTCASE(listnet_ranker, generated)
+{
+    Xapian::ListNETRanker ranker;
+    TEST_EXCEPTION(Xapian::FileNotFoundError, ranker.train_model(""));
+    string db_path = get_database_path("apitest_ranker1",
+				       db_index_two_documents);
+    Xapian::Enquire enquire((Xapian::Database(db_path)));
+    enquire.set_query(Xapian::Query("lions"));
+    Xapian::MSet mymset = enquire.get_mset(0, 10);
+    std::string data_directory = test_driver::get_srcdir() + "/testdata/";
+    Xapian::prepare_training_file(db_path, data_directory + "query.txt",
+				  data_directory + "qrel.txt", 10,
+				  "listnet_ranker_training.txt");
+    ranker.set_database_path(db_path);
+    TEST_EQUAL(ranker.get_database_path(), db_path);
+    ranker.set_query(Xapian::Query("lions"));
+    ranker.train_model("listnet_ranker_training.txt");
+    Xapian::docid doc1 = *mymset[0];
+    Xapian::docid doc2 = *mymset[1];
+    ranker.rank(mymset);
+    TEST_EQUAL(doc2, *mymset[0]);
+    TEST_EQUAL(doc1, *mymset[1]);
+    mymset = enquire.get_mset(0, 10);
+    ranker.train_model("listnet_ranker_training.txt", "ListNet_Ranker");
+    ranker.rank(mymset, "ListNet_Ranker");
+    TEST_EQUAL(doc2, *mymset[0]);
+    TEST_EQUAL(doc1, *mymset[1]);
+    TEST_EXCEPTION(Xapian::LetorInternalError,
+		   ranker.score(data_directory + "query.txt", data_directory +
+				"qrel.txt", "ListNet_Ranker",
+				"scorer_output.txt", 10, ""));
+    TEST_EXCEPTION(Xapian::FileNotFoundError,
+		   ranker.score("", data_directory + "qrel.txt",
+				"ListNet_Ranker", "scorer_output.txt", 10));
+    TEST_EXCEPTION(Xapian::FileNotFoundError,
+		   ranker.score(data_directory + "qrel.txt", "",
+				"ListNet_Ranker", "scorer_output.txt", 10));
+    ranker.score(data_directory + "query.txt", data_directory + "qrel.txt",
+		 "ListNet_Ranker", "scorer_output.txt", 10);
```

Comment:
I'd suggest just creating the two filenames once, e.g.:

```
string data_directory = test_driver::get_srcdir() + "/testdata/";
string query_txt = data_directory + "query.txt";
string qrel_txt = data_directory + "qrel.txt";
```

And then you can just use `query_txt` and `qrel_txt` everywhere.  That'll make the code a bit more readable, and probably be more efficient (because the compiler probably isn't smart enough to hoist a common subexpression of type `std::string`).